src/components/DecisioningEngine/createEventRegistry.js

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2023 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+import { createRestoreStorage, createSaveStorage } from "./utils";
+
+const STORAGE_KEY = "events";
+const MAX_EVENT_RECORDS = 1000;
+const DEFAULT_SAVE_DELAY = 500;
+
+export const createEventPruner = (limit = MAX_EVENT_RECORDS) => {
+  return events => {
+    const pruned = {};
+    Object.keys(events).forEach(eventType => {
+      pruned[eventType] = {};
+      Object.values(events[eventType])
+        .sort((a, b) => a.firstTimestamp - b.firstTimestamp)
```

Comment:
How would i know which one is the oldest one without sorting?  These events are keyed by id.

I was worried about this running too frequently too.  My thought was to leave it and perhaps only do the pruning when initialized.  That way more than 1000 events could be in the storage but will be trimmed back on next page load.  What do you think?

I'm not sure I follow your big O formula.  This is literally iterating across only the relevant entries and transferring them to a new object for storage.  O(N).  I guess the sort isn't for free, but can you explain your formula and what you are trying to say in more detail please?

Reply:
Sort is O(M log M) where M is the number of items. Finding the smallest timestamp is O(M). If you only ever have to remove one item, then you can get away with only finding the smallest. I guess it doesn't matter much when we are talking about these small numbers and session storage is limited in size anyways, but if it doesn't matter, then let's take out the debounce so that we get all the events.

Reply:
Sure, i can remove the debounce if you don't think it adds value.  Seems to me that debouncing is worthwhile though.  At least somewhere in the process.  Consider if a customer calls `alloy('renderDecisions')` on scroll every tick...

Reply:
I see. I was just considering events that get sent to experience edge. Perhaps when there is an actual sendEvent we would not do the debounce? I'm not sure how we would do that in the code though.